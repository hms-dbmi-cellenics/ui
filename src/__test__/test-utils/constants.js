export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
