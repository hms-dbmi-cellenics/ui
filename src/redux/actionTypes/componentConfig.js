/**
 * Sets plot configs and data.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const RESET_CONFIG = `${COMPONENT_CONFIG}/reset`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;
const PLOT_DATA_LOADED = `${COMPONENT_CONFIG}/plotDataLoaded`;
const PLOT_DATA_LOADING = `${COMPONENT_CONFIG}/plotDataLoading`;
const PLOT_DATA_ERROR = `${COMPONENT_CONFIG}/plotDataError`;

export {
  LOAD_CONFIG, UPDATE_CONFIG, RESET_CONFIG, SAVE_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR,
};
