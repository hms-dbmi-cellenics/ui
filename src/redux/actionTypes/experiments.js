const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

const EXPERIMENTS_SWITCH = `${EXPERIMENTS}/switch`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SWITCH,
};
