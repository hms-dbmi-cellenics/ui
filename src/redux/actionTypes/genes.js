const GENES = 'genes';

/**
 * Turns on the loading condition for gene information.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const GENES_PROPERTIES_LOADING = `${GENES}/propertiesLoading`;

/**
 * Sets the state of the gene data to be successfully loaded, with the appropriate gene data.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const GENES_PROPERTIES_LOADED_PAGINATED = `${GENES}/propertiesLoadedPaginated`;

/**
 * Sets an error condition for gene properties data.
 */
const GENES_PROPERTIES_ERROR = `${GENES}/propertiesError`;

/**
 * Updates the list of selected genes with a new selected gene.
 */
const GENES_SELECT = `${GENES}/select`;

/**
 * Removes a deselected gene from the list of selected genes.
 */
const GENES_DESELECT = `${GENES}/deselect`;

/**
 * Turns on the loading state for expression data.
 */
const GENES_EXPRESSION_LOADING = `${GENES}/expressionLoading`;

/**
 * Sets the state of the expression store to be successfully loaded,
 * with the appropriate expression data.
 */
const GENES_EXPRESSION_LOADED = `${GENES}/expressionLoaded`;

/**
 * Sets an error condition for gene expression data.
 */
const GENES_EXPRESSION_ERROR = `${GENES}/expressionError`;

/**
 * Sets marker genes as loading
 */
const MARKER_GENES_LOADING = `${GENES}/markerGenesLoading`;

/**
 * Marker genes have finished loading
 */
const MARKER_GENES_LOADED = `${GENES}/markerGenesLoaded`;

/**
 * Marker genes load task had some kind of error
 */
const MARKER_GENES_ERROR = `${GENES}/markerGenesError`;

/**
 * Downsampled genes expression load task
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADING = `${GENES}/downsampledGenesExpressionLoading`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADED = `${GENES}/downsampledGenesExpressionLoaded`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_ERROR = `${GENES}/downsampledGenesExpressionError`;

export {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  DOWNSAMPLED_GENES_EXPRESSION_LOADING,
  DOWNSAMPLED_GENES_EXPRESSION_LOADED,
  DOWNSAMPLED_GENES_EXPRESSION_ERROR,
};
