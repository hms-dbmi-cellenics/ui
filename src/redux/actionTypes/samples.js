const SAMPLES = 'samples';

/**
 * Save sample.
 */
const SAMPLES_CREATE = `${SAMPLES}/create`;

/**
 * Update sample.
 */
const SAMPLES_UPDATE = `${SAMPLES}/update`;

/**
 * Delete sample.
 */
const SAMPLES_DELETE_API_V2 = `${SAMPLES}/deleteApiV2`;

/**
 * Delete sample following apiv1 version, remove when doing the final step of cleaning up v1 code.
 */
const SAMPLES_DELETE_API_V1 = `${SAMPLES}/deleteApiV1`;

/**
 * Update files in sample.
 */
const SAMPLES_FILE_UPDATE = `${SAMPLES}/fileUpdate`;

/**
 * Load sample.
 */
const SAMPLES_LOADED = `${SAMPLES}/loaded`;

/**
 * Saving samples.
 */
const SAMPLES_SAVING = `${SAMPLES}/saving`;

/**
 * Samples saved successfully.
 */
const SAMPLES_SAVED = `${SAMPLES}/saved`;

/**
 * Error loading/saving sample.
 */
const SAMPLES_ERROR = `${SAMPLES}/error`;

/**
 * Delete metada from sample.
 */
const SAMPLES_METADATA_DELETE = `${SAMPLES}/metadataDelete`;

const SAMPLES_LOADING = `${SAMPLES}/loading`;
export {
  SAMPLES_CREATE,
  SAMPLES_UPDATE,
  SAMPLES_DELETE_API_V2,
  SAMPLES_DELETE_API_V1,
  SAMPLES_FILE_UPDATE,
  SAMPLES_LOADED,
  SAMPLES_SAVING,
  SAMPLES_ERROR,
  SAMPLES_SAVED,
  SAMPLES_METADATA_DELETE,
  SAMPLES_LOADING,
};
