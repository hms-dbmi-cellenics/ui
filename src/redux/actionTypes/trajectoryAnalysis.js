const TRAJECTORY_ANALYSIS = 'trajectoryAnalysis';

/**
 * Loading trajectory analysis data
 */
const TRAJECTORY_ANALYSIS_LOADING = `${TRAJECTORY_ANALYSIS}/loading`;

/**
 * Trajectory analysis data loaded
 */
const TRAJECTORY_ANALYSIS_LOADED = `${TRAJECTORY_ANALYSIS}/loaded`;

/**
 * Error loading data
 */
const TRAJECTORY_ANALYSIS_ERROR = `${TRAJECTORY_ANALYSIS}/error`;

export {
  TRAJECTORY_ANALYSIS_LOADING,
  TRAJECTORY_ANALYSIS_LOADED,
  TRAJECTORY_ANALYSIS_ERROR,
};
