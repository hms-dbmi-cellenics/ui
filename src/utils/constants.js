import endUserMessages from 'utils/endUserMessages';

const api = {
  V1: 'v1',
  V2: 'v2',
};

const modules = {
  DATA_MANAGEMENT: 'DATA_MANAGEMENT',
  REPOSITORY: 'REPOSITORY',
  DATA_PROCESSING: 'DATA_PROCESSING',
  DATA_EXPLORATION: 'DATA_EXPLORATION',
  PLOTS_AND_TABLES: 'PLOTS_AND_TABLES',
  SETTINGS: 'SETTINGS',
  DEFAULT: '',
};

const sampleTech = {
  '10X': '10x',
  RHAPSODY: 'rhapsody',
  SEURAT: 'seurat',
  H5: '10x_h5',
};

const plotTypes = {
  CONTINUOUS_EMBEDDING: 'embeddingContinuous',
  CATEGORICAL_EMBEDDING: 'embeddingCategorical',
  MARKER_HEATMAP: 'markerHeatmap',
  VOLCANO_PLOT: 'volcano',
  FREQUENCY_PLOT: 'frequency',
  VIOLIN_PLOT: 'violin',
  DOT_PLOT: 'DotPlot',
  TRAJECTORY_ANALYSIS: 'TrajectoryAnalysis',
  NORMALIZED_EXPRESSION_MATRIX: 'NormalizedExpressionMatrix',
  MULTI_VIEW_PLOT: 'multiView',
};

const plotUuids = {
  CONTINUOUS_EMBEDDING: 'embeddingContinuousMain',
  CATEGORICAL_EMBEDDING: 'embeddingCategoricalMain',
  MARKER_HEATMAP: 'markerHeatmapPlotMain',
  VOLCANO_PLOT: 'volcanoPlotMain',
  FREQUENCY_PLOT: 'frequencyPlotMain',
  VIOLIN_PLOT: 'ViolinMain',
  DOT_PLOT: 'DotPlotMain',
  TRAJECTORY_ANALYSIS: 'trajectoryAnalysisMain',
  NORMALIZED_EXPRESSION_MATRIX: 'normalized-matrix',
  getMultiPlotUuid: (plotType) => `${plotTypes.MULTI_VIEW_PLOT}-${plotType}`,
};

const plotNames = {
  CONTINUOUS_EMBEDDING: 'Continuous Embedding',
  CATEGORICAL_EMBEDDING: 'Categorical Embedding',
  MARKER_HEATMAP: 'Heatmap',
  VOLCANO_PLOT: 'Volcano Plot',
  FREQUENCY_PLOT: 'Frequency Plot',
  VIOLIN_PLOT: 'Violin Plot',
  DOT_PLOT: 'Dot Plot',
  TRAJECTORY_ANALYSIS: 'Trajectory Analysis',
  NORMALIZED_EXPRESSION_MATRIX: 'Normalized Expression Matrix',
  BATCH_DIFFERENTIAL_EXPRESSION: 'Batch Differential Expression Table',
};

const layout = {
  PANEL_HEADING_HEIGHT: 30,
  PANEL_PADDING: 10,
};

const downsamplingMethods = {
  GEOSKETCH: 'geosketch',
  NONE: 'none',
  DEFAULT_PERC_TO_KEEP: 5,
};

const cellSetsUpdatedMessages = {
  ClusterCells: endUserMessages.SUCCESS_CELL_SETS_RECLUSTERED,
  ScTypeAnnotate: endUserMessages.SUCCESS_CELL_SETS_ANNOTATED,
};

export {
  api,
  modules,
  sampleTech,
  plotTypes,
  plotUuids,
  plotNames,
  layout,
  downsamplingMethods,
  cellSetsUpdatedMessages,
};
