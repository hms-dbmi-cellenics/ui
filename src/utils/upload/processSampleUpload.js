/* eslint-disable no-param-reassign */
import _ from 'lodash';

import {
  createSamples, createSampleFile, updateSampleFileUpload, validateSamples,
} from 'redux/actions/samples';

import UploadStatus from 'utils/upload/UploadStatus';
import { inspectFile, Verdict } from 'utils/upload/fileInspector';
import fetchAPI from 'utils/http/fetchAPI';

import { sampleTech } from 'utils/constants';
import fileUploadUtils from 'utils/upload/fileUploadUtils';
import endUserMessages from 'utils/endUserMessages';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import uploadFileToS3 from 'utils/upload/multipartUpload';

const createAndUploadSampleFile = async (
  file, fileType, experimentId, sampleId, dispatch, selectedTech,
) => {
  const abortController = new AbortController();

  let sampleFileId;

  try {
    sampleFileId = await dispatch(
      createSampleFile(
        experimentId,
        sampleId,
        fileType,
        file,
        abortController,
      ),
    );
  } catch (e) {
    pushNotificationMessage('error', endUserMessages.ERROR_BEGIN_SAMPLE_FILE_UPLOAD);
    return;
  }

  // Take the fileName now because after loadAndCompressIfNecessary the name could
  // be lost in the compression. If it is compressed fileObject becomes a uInt8Array
  // instead of the fileReader metadata object that it is now
  const fileName = file.fileObject.name;

  try {
    const { uploadId, bucket, key } = await beginSampleFileUpload(
      experimentId,
      sampleFileId,
      getMetadata(fileName, selectedTech),
    );

    const updateSampleFileUploadProgress = (status, percentProgress = 0) => {
      dispatch(
        updateSampleFileUpload(
          experimentId, sampleId, sampleFileId, fileType, status, percentProgress,
        ),
      );
    };

    const uploadUrlParams = {
      uploadId, fileId: sampleFileId, bucket, key,
    };

    await uploadFileToS3(
      experimentId,
      file,
      !file.compressed,
      uploadUrlParams,
      'sample',
      abortController,
      updateSampleFileUploadProgress,
    );
  } catch (e) {
    dispatch(updateSampleFileUpload(
      experimentId, sampleId, sampleFileId, fileType, UploadStatus.UPLOAD_ERROR,
    ));
  }
};

const beginSampleFileUpload = async (experimentId, sampleFileId, metadata) => await fetchAPI(
  `/v2/experiments/${experimentId}/sampleFiles/${sampleFileId}/beginUpload`,
  {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ metadata }),
  },
);

const getMetadata = (fileName, selectedTech) => {
  const metadata = {};
  if (selectedTech === sampleTech['10X']) {
    if (fileName.includes('genes')) {
      metadata.cellranger_version = 'v2';
    } else if (fileName.includes('features')) {
      metadata.cellranger_version = 'v3';
    }
  }

  return metadata;
};

const processSampleUpload = async (filesList, technology, samples, experimentId, dispatch) => {
  // First use map to make it easy to add files in the already existing sample entry
  const samplesMap = filesList.reduce((acc, file) => {
    const { sample: sampleName, name } = fileUploadUtils[technology].getFileSampleAndName(file.fileObject.path.replace(/[\s]{2,}/ig, ' '));

    const fileType = fileUploadUtils[technology].getCorrespondingType(name);

    const sampleUuid = Object.values(samples).filter(
      (s) => s.name === sampleName
        && s.experimentId === experimentId,
    )[0]?.uuid;

    return {
      ...acc,
      [sampleName]: {
        ...acc[sampleName],
        uuid: sampleUuid,
        files: {
          ...acc[sampleName]?.files,
          [fileType]: file,
        },
      },
    };
  }, {});

  const validSamplesList = await dispatch(validateSamples(experimentId, samplesMap, technology));

  // If none of the files are in valid format, return
  if (validSamplesList.length === 0) return;

  // Sort alphabetically
  validSamplesList.sort(([oneName], [otherName]) => oneName.localeCompare(otherName));

  try {
    const sampleIdsByName = await dispatch(
      createSamples(
        experimentId,
        validSamplesList,
        technology,
      ),
    );

    const promises = [];

    validSamplesList.forEach(([name, sample]) => {
      Object.entries(sample.files).forEach(([type, file]) => {
        promises.push(
          async () => await createAndUploadSampleFile(
            file,
            type,
            experimentId,
            sampleIdsByName[name],
            dispatch,
            technology,
          ),
        );
      });
    });

    // 5 at a time
    const chunkedPromises = _.chunk(promises, promises.length);

    // eslint-disable-next-line no-restricted-syntax
    for await (const promisesChunk of chunkedPromises) {
      await Promise.all(promisesChunk.map((promise) => promise()));
    }
  } catch (e) {
    // Ignore the error, if createSamples fails we throw to
    // avoid attempting to upload any of these broken samples
  }
};

/**
   * This function converts an uploaded File object into a file record that will be inserted under
   * samples[files] in the redux store.
   * @param {File} fileObject the File object that is uploaded to the .
   * @param {string} technology the chosen technology that outputs this file. Used for verification.
   * @returns {object} fileRecord object that will be associated with a sample.
   */
const fileObjectToFileRecord = async (fileObject, technology) => {
  // This is the first stage in uploading a file.

  const verdict = await inspectFile(fileObject, technology);

  let error = '';
  if (verdict === Verdict.INVALID_NAME) {
    error = 'Invalid file name.';
  } else if (verdict === Verdict.INVALID_FORMAT) {
    error = 'Invalid file format.';
  }

  return {
    fileObject,
    size: fileObject.size,
    path: fileObject.path,
    upload: {
      status: UploadStatus.UPLOADING,
      progress: 0,
    },
    errors: error,
    compressed: verdict === Verdict.VALID_ZIPPED,
  };
};

export {
  fileObjectToFileRecord,
  createAndUploadSampleFile,
};

export default processSampleUpload;
