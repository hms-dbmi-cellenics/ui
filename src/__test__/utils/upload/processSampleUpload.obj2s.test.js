import _ from 'lodash';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import waitForActions from 'redux-mock-store-await-actions';
import axios from 'axios';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { sampleTech } from 'utils/constants';
import { waitFor } from '@testing-library/react';

import { SAMPLES_FILE_UPDATE } from 'redux/actionTypes/samples';
import initialSampleState, { sampleTemplate } from 'redux/reducers/samples/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

import UploadStatus from 'utils/upload/UploadStatus';

import processSampleUpload from 'utils/upload/processSampleUpload';

import validate from 'utils/upload/validateObj2s';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import mockFile from '__test__/test-utils/mockFile';

const MB = 1024 * 1024;
jest.mock('../../../utils/upload/fileUploadConfig', () => ({
  chunkSize: 1 * MB,
}));

enableFetchMocks();

const getValidFiles = (fileSize = 1 * MB) => {
  const obj2sFiles = [{
    name: 'r.rds',
    fileObject: mockFile('r.rds', '', fileSize),
    size: fileSize,
    upload: { status: UploadStatus.UPLOADING },
    errors: '',
    compressed: true,
    valid: true,
  }];

  return obj2sFiles;
};

const sampleType = sampleTech.SEURAT;
const mockSampleUuid = 'sample-uuid';
const mockExperimentId = 'project-uuid';
const sampleName = 'mockSampleName';

const initialState = {
  experiments: {
    ...initialExperimentState,
    [mockExperimentId]: {
      ...experimentTemplate,
      id: mockExperimentId,
    },
  },
  samples: {
    ...initialSampleState,
    ids: [mockSampleUuid],
    meta: {
      loading: true,
      error: false,
    },
    [mockSampleUuid]: {
      ...sampleTemplate,
      uuid: mockSampleUuid,
      name: sampleName,
      experimentId: mockExperimentId,
    },
  },
};

const mockStore = configureMockStore([thunk]);

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'sample-uuid'),
}));

jest.mock('axios', () => ({
  request: jest.fn(),
}));

jest.mock('utils/pushNotificationMessage');
jest.mock('utils/upload/validateObj2s');

let store = null;

describe('processUpload', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    const uploadId = 'some_id';
    const bucket = 'biomage-originals-test-accountId';
    const mockSampleFileId = 'mockSampleFileId';

    const mockUploadUrlParams = {
      uploadId,
      bucket,
      key: mockSampleFileId,
    };

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockIf(/.*/, ({ url }) => {
      let result;

      if (url.endsWith(`/v2/experiments/${mockExperimentId}/samples`)) {
        result = {
          status: 200,
          body: JSON.stringify([]),
        };
      }

      // Create sample file
      if (new RegExp(`/v2/experiments/${mockExperimentId}/samples/.*/sampleFiles/.*`).test(url)) {
        result = { status: 200, body: JSON.stringify({}) };
      }

      // Update sample file status
      if (new RegExp(`/v2/experiments/${mockExperimentId}/sampleFiles/.*`).test(url)) {
        result = { status: 200, body: JSON.stringify({}) };
      }

      if (url.endsWith(`/v2/experiments/${mockExperimentId}/sampleFiles/${mockSampleUuid}/beginUpload`)) {
        result = { status: 200, body: JSON.stringify(mockUploadUrlParams) };
      }

      const queryParams = new URLSearchParams({ bucket, key: mockSampleFileId });
      if (url.endsWith(`/v2/experiments/${mockExperimentId}/upload/${uploadId}/part/1/signedUrl?${queryParams}`)) {
        result = { status: 200, body: JSON.stringify('theSignedUrl') };
      }

      if (url.endsWith(`/v2/experiments/${mockExperimentId}/upload/${uploadId}/part/2/signedUrl?${queryParams}`)) {
        result = { status: 200, body: JSON.stringify('theSignedUrl2') };
      }

      if (url.endsWith('/v2/completeMultipartUpload')) {
        result = { status: 200, body: JSON.stringify({}) };
      }

      return Promise.resolve(result);
    });

    store = mockStore(initialState);
  });

  it('Uploads and updates redux correctly when there are no errors', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementation(uploadSuccess);

    await processSampleUpload(
      getValidFiles(),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(1).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADED } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Two axios put calls are made (one for each presigned url)
    expect(mockAxiosCalls.length).toBe(1);
    // Each put call is made with the correct information
    expect(mockAxiosCalls[0].url).toBe('theSignedUrl');

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );
    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);
    const uploadingStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const uploadedStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 1 files actions with status uploading
    expect(uploadingStatusProperties.length).toEqual(1);
    // There are 1 files actions with status uploaded
    expect(uploadedStatusProperties.length).toEqual(1);

    // axios request calls are correct
    expect(axios.request.mock.calls.map((call) => _.omit(call[0], 'data'))).toMatchSnapshot();

    // If we trigger axios onUploadProgress for the two presigned urls,
    // it updates the progress correctly
    axios.request.mock.calls[0][0].onUploadProgress({ progress: 0.25 });

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADING, progress: 25 } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );
  });

  it('Updates redux correctly when there are file upload errors', async () => {
    const mockAxiosCalls = [];

    const uploadError = (params) => {
      mockAxiosCalls.push(params);
      return Promise.reject(new Error('Error'));
    };

    axios.request.mockImplementation(uploadError);

    await processSampleUpload(
      getValidFiles(),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(1).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOAD_ERROR } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );

    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);

    const uploadingFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );

    const errorFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOAD_ERROR,
    );

    const uploadedFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 1 files actions with status uploading
    expect(uploadingFileProperties.length).toEqual(1);
    // There are 1 files actions with status upload error
    expect(errorFileProperties.length).toEqual(1);
    // There are no file actions with status successfully uploaded
    expect(uploadedFileProperties.length).toEqual(0);
  });

  it('Should not upload files if there are errors creating samples in the api', async () => {
    fetchMock.mockReject(new Error('Error'));

    await processSampleUpload(
      getValidFiles(),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We do not expect uploads to happen
    await waitFor(() => {
      expect(axios.request).not.toHaveBeenCalled();
    });
  });

  it('Should not validate .rds files', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementation(uploadSuccess);

    validate.mockImplementationOnce(
      () => (['Some file error']),
    );

    await processSampleUpload(
      getValidFiles(),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We expect uploads to happen
    await waitFor(() => {
      expect();
      expect(pushNotificationMessage).toHaveBeenCalledTimes(0);

      // 1 part
      expect(axios.request).toHaveBeenCalledTimes(1);
      expect(validate).toHaveBeenCalledTimes(1);
    });
  });

  it('Should be able to upload files that require 2 parts', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementation(uploadSuccess);

    await processSampleUpload(
      getValidFiles(2 * MB),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We expect uploads to happen
    await waitFor(() => {
      expect(pushNotificationMessage).toHaveBeenCalledTimes(0);

      // 2 parts
      expect(axios.request).toHaveBeenCalledTimes(2);
      expect(validate).toHaveBeenCalledTimes(1);
    });
  });
});
