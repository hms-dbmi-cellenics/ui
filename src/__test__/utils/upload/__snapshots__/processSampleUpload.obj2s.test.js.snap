// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processUpload Uploads and updates redux correctly when there are no errors 1`] = `
[
  {
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "signal": AbortSignal {},
    "url": "theSignedUrl",
  },
]
`;
